### PR TITLE
Allow user to opt-in to the different client policies

### PR DESCRIPTION
**Pull request by PatoBeltran**
## Bug
Third part of NuGet/Home#6961

## Fix
When creating the `SignedPackageVerifierSettings` to give to the `ExtractionContext` we need to read the client policy information from the `ISettings` and set it properly to then consume it in the signing verification API. 

This PR gives the ability to the user to add `trustedSigners` to the nuget.config and have a more strict validation.

This PR is missing to honor the `allowUntrustedRoot` attribute from a certificate item. This will come in a subsequent PR.

### CODE REVIEW

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs [hunk 1]
```diff
@@ -77,6 +64,21 @@ public sealed class VsPathContextProvider : IVsPathContextProvider2
             _solutionManager = solutionManager ?? throw new ArgumentNullException(nameof(solutionManager));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
             _getLockFileOrNullAsync = BuildIntegratedProjectUtility.GetLockFileOrNull;
+
+            _projectContext = new Lazy<INuGetProjectContext>(() => {
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs:68]: why is this changed?

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/AllowListVerificationProvider.cs [hunk 2]
```diff
@@ -106,6 +106,11 @@ private Task<SignedPackageVerificationResult> VerifyAllowList(CertificateListVer
                         if (IsSignatureTargeted(certificateHashEntry.Target, signature) &&
                             StringComparer.OrdinalIgnoreCase.Equals(certificateHashEntry.Fingerprint, primarySignatureCertificateFingerprint))
                         {
+                            if (ShouldVerifyOwners(certificateHashEntry as TrustedRepositoryAllowListEntry, signature as IRepositorySignature, out var allowedOwners, out var actualOwners))
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/AllowListVerificationProvider.cs]: `certificateHashEntry as TrustedRepositoryAllowListEntry, signature as IRepositorySignature` could this ever throw? also personally never liked multiple `out` arguments. It always makes it harder to read.

File: src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+using System.Linq;
+using NuGet.Configuration;
+
+namespace NuGet.Packaging.Signing
+{
+    public class TrustedSignersProvider
+    {
+        private readonly ISettings _settings;
+
+        public TrustedSignersProvider(ISettings settings)
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs]: since you immediately call `GetAllowListEntries` after constructing this object so instead of persisting `ISettings` you can simply pass it in `GetAllowListEntries` so that in future, it's able to return the new entries when underneath Settings are changed dynamically through UI.

**Review by jainaashish — APPROVED** [review]
Looks good to me. Couple of small suggestions.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs (hunk 1 above)
Comment (PatoBeltran) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs:68]: because I could not pass `_settings.Value` in `GetClientPolicy(...)` if the creation of the lazy object was not inside of a method, therefore I moved it to the constructor.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/AllowListVerificationProvider.cs (hunk 2 above)
Comment (PatoBeltran) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/AllowListVerificationProvider.cs]: This will never throw since if it's not of that type it will return `null`. Do you think it would be more readable to have the code inline instead of a separate function and have it repeated in both places it needed to avoid the multiple `out` arguments?

File: src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+using System.Globalization;
+using System.Linq;
+using NuGet.Common;
+using NuGet.Configuration;
+
+namespace NuGet.Packaging.Signing
+{
+    public static class TrustedSignersProvider
+    {
+        public static IReadOnlyList<VerificationAllowListEntry> GetAllowListEntries(ISettings settings, ILogger logger)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs:15]: Should this be async maybe?

File: src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                            }
+                            else
+                            {
+                                existingEntry.Owners.AddRange(owners);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        certificateLookup.Add($"{certificate.HashAlgorithm.ToString()}-{certificate.Fingerprint}", new CertificateEntryLookupEntry(itemTarget, itemPlacement, certificate));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs]: nit: move the key generation in a common method to prevent bad typos.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/CertificateHashAllowListEntry.cs [hunk 6]
```diff
@@ -14,8 +17,65 @@ public class CertificateHashAllowListEntry : VerificationAllowListEntry
         public CertificateHashAllowListEntry(VerificationTarget target, SignaturePlacement placement, string fingerprint, HashAlgorithmName algorithm)
             : base(target, placement)
         {
-            Fingerprint = fingerprint;
+            if (!Enum.IsDefined(typeof(SignaturePlacement), placement))
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/CertificateHashAllowListEntry.cs:20]: why?
How can this happen?
There's no serialization that could break it.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignedPackageVerifierSettings.cs [hunk 7]
```diff
@@ -221,6 +221,29 @@ public sealed class SignedPackageVerifierSettings
             ClientCertificateList = clientAllowListEntries;
         }
 
+        /// <summary>
+        /// Gives the appropriate configuration depending on the user specified settings.
+        /// </summary>
+        /// <param name="settings">Loaded settings used to get user data</param>
+        public static SignedPackageVerifierSettings GetClientPolicy(ISettings settings, ILogger logger)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignedPackageVerifierSettings.cs:228]: Maybe this can be async.

File: src/NuGet.Core/NuGet.Packaging/Strings.resx [hunk 8]
```diff
@@ -747,4 +747,9 @@ Valid from:</comment>
   <data name="NuGetLicenseExpression_LicenseInvalidCharacters" xml:space="preserve">
     <value>The license identifier '{0}' contains invalid characters.</value>
   </data>
+  <data name="ConflictingAllowUntrustedRoot" xml:space="preserve">
+    <value>There are two certificates with conflicting allowUntrustedRoot in the computed settings. AllowUntrustedRoot is going to be set to false. Certificate: {0}-{1}</value>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Strings.resx]: nit: I get it's the beginning of a sentence but maybe the AllowUntrustedRoot casing should be consistent. 

The real feedbackh ere is to add a comment specifying that allowUntrustedRoot should not be localized.

**Review by nkolev92 — APPROVED** [review]
A few nits and comments. 

Just wondering if some of these processing methods should be made async.

File: src/NuGet.Core/NuGet.Packaging/Strings.resx (hunk 8 above)
Comment (PatoBeltran) [reply src/NuGet.Core/NuGet.Packaging/Strings.resx]: I used that casing since `allowUntrustedRoot` uses that casing in the config file. I will add the comment to not localize it. Good catch!

File: src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs (hunk 4 above)
Comment (PatoBeltran) [reply src/NuGet.Core/NuGet.Packaging/Signing/TrustedSigners/TrustedSignersProvider.cs:15]: None of the `ISettings` operations are async. We could discuss if we want to add async support for `ISettings`... Since it is a read operation is working in the computed settings, therefore there is no need for it to be async.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/CertificateHashAllowListEntry.cs (hunk 6 above)
Comment (PatoBeltran) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/CertificateHashAllowListEntry.cs:20]: There is no case where this will break, I added this for the sake of the sanity of the API itself, in case there is an external user to not allow it to break. Do you think it should be deleted?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/CertificateHashAllowListEntry.cs:20]: No, not necessarily, it doesn't hurt, but I also don't think it has to be there. 

We've normally only done this in plugin code, but if you feel strongly about it, keep it :)

Comment (PatoBeltran) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/CertificateHashAllowListEntry.cs:20]: We also have this in other verification settings, which is where I got it from.
